In [1]:
import pandas as pd
dataset=pd.read_csv(r"metadata (2).csv")
dataset

rarespecies_id  eol_content_id  eol_page_id  \
0      dc48f2ce-4feb-4ef7-b2a2-c3c3f42bf19b        28657539       491832   
1      7faca96a-54e6-4c80-b9e4-77ab126d904a        21232818      1033999   
2      16b767da-7b35-4f76-8ef7-0835148b760f        29911923       132324   
3      32e472b8-722a-4983-8fe5-a0605e330c1b        29856158       485420   
4      81e2349d-0634-4f34-8b31-384f1d1d81f2        29372149       354871   
...                                     ...             ...          ...   
11978  c74baf60-4a1d-4845-bce6-ba486548766d         2751830     46473744   
11979  b8c44794-14bf-4a3d-ad7f-cd4ea00a21e7        22914588     46473747   
11980  34be9673-fa25-4d7d-9022-2740b6f34fcb        20298397      4792502   
11981  a629ccce-bda5-486a-a39f-b805ec70fd52        21566265       449459   
11982  f56af9c5-72b9-4252-9e61-00e32bf08c91        12835846     46454284   

        kingdom       class          order          family        genus  \
0      Animalia     Insecta    Hymenoptera      Formicidae   Adetomyrma   
1      Animalia     Insecta        Odonata       Gomphidae  Hemigomphus   
2      Animalia     Insecta        Odonata  Platystictidae  Protosticta   
3      Animalia     Insecta    Hymenoptera      Formicidae     Pheidole   
4      Animalia     Insecta     Coleoptera    Cerambycidae  Macrodontia   
...         ...         ...            ...             ...          ...   
11978  Animalia    Bivalvia       Cardiida       Cardiidae     Tridacna   
11979  Animalia    Bivalvia       Cardiida       Cardiidae     Tridacna   
11980  Animalia  Gastropoda    Lepetellida      Haliotidae     Haliotis   
11981  Animalia    Bivalvia       Unionida       Unionidae    Potamilus   
11982  Animalia  Gastropoda  Neogastropoda         Conidae        Conus   

              species                     sciName               common  \
0            venatrix         Adetomyrma venatrix          dracula ant   
1       theischingeri   Hemigomphus theischingeri  Rainforest Vicetail   
2      sanguinostigma  Protosticta sanguinostigma    red spot reedtail   
3            elecebra           Pheidole elecebra                  ant   
4         cervicornis     Macrodontia cervicornis   Giant Jawed Sawyer   
...               ...                         ...                  ...   
11978           gigas              Tridacna gigas     Gigas Giant Clam   
11979        squamosa           Tridacna squamosa    Fluted Giant Clam   
11980            roei               Haliotis roei        Roe's abalone   
11981    amphichaenus      Potamilus amphichaenus   Texas heelsplitter   
11982        mercator              Conus mercator        Merchant Cone   

                                               file_name      phylum  
0      dataset/Animalia-Arthropoda-Insecta-Hymenopter...  Arthropoda  
1      dataset/Animalia-Arthropoda-Insecta-Odonata-Go...  Arthropoda  
2      dataset/Animalia-Arthropoda-Insecta-Odonata-Pl...  Arthropoda  
3      dataset/Animalia-Arthropoda-Insecta-Hymenopter...  Arthropoda  
4      dataset/Animalia-Arthropoda-Insecta-Coleoptera...  Arthropoda  
...                                                  ...         ...  
11978  dataset/Animalia-Mollusca-Bivalvia-Cardiida-Ca...    Mollusca  
11979  dataset/Animalia-Mollusca-Bivalvia-Cardiida-Ca...    Mollusca  
11980  dataset/Animalia-Mollusca-Gastropoda-Lepetelli...    Mollusca  
11981  dataset/Animalia-Mollusca-Bivalvia-Unionida-Un...    Mollusca  
11982  dataset/Animalia-Mollusca-Gastropoda-Neogastro...    Mollusca  

[11983 rows x 13 columns]

In [2]:
#deleting columns that are not needed
dataset.drop(['rarespecies_id','eol_content_id','eol_page_id','kingdom','sciName','common','file_name'],axis=1, inplace=True)
dataset

class          order          family        genus         species  \
0         Insecta    Hymenoptera      Formicidae   Adetomyrma        venatrix   
1         Insecta        Odonata       Gomphidae  Hemigomphus   theischingeri   
2         Insecta        Odonata  Platystictidae  Protosticta  sanguinostigma   
3         Insecta    Hymenoptera      Formicidae     Pheidole        elecebra   
4         Insecta     Coleoptera    Cerambycidae  Macrodontia     cervicornis   
...           ...            ...             ...          ...             ...   
11978    Bivalvia       Cardiida       Cardiidae     Tridacna           gigas   
11979    Bivalvia       Cardiida       Cardiidae     Tridacna        squamosa   
11980  Gastropoda    Lepetellida      Haliotidae     Haliotis            roei   
11981    Bivalvia       Unionida       Unionidae    Potamilus    amphichaenus   
11982  Gastropoda  Neogastropoda         Conidae        Conus        mercator   

           phylum  
0      Arthropoda  
1      Arthropoda  
2      Arthropoda  
3      Arthropoda  
4      Arthropoda  
...           ...  
11978    Mollusca  
11979    Mollusca  
11980    Mollusca  
11981    Mollusca  
11982    Mollusca  

[11983 rows x 6 columns]

In [3]:
#deleting duplicates directly on the original dataset without creating a new one
dataset.drop_duplicates(inplace=True)
dataset

class            order          family        genus  \
0         Insecta      Hymenoptera      Formicidae   Adetomyrma   
1         Insecta          Odonata       Gomphidae  Hemigomphus   
2         Insecta          Odonata  Platystictidae  Protosticta   
3         Insecta      Hymenoptera      Formicidae     Pheidole   
4         Insecta       Coleoptera    Cerambycidae  Macrodontia   
...           ...              ...             ...          ...   
11775  Gastropoda    Neogastropoda         Conidae        Conus   
11776  Gastropoda      Lepetellida      Haliotidae     Haliotis   
11777    Bivalvia         Cardiida       Cardiidae     Tridacna   
11788  Gastropoda  Stylommatophora       Zonitidae      Zonites   
11798    Bivalvia         Cardiida       Cardiidae     Tridacna   

              species      phylum  
0            venatrix  Arthropoda  
1       theischingeri  Arthropoda  
2      sanguinostigma  Arthropoda  
3            elecebra  Arthropoda  
4         cervicornis  Arthropoda  
...               ...         ...  
11775        mercator    Mollusca  
11776            roei    Mollusca  
11777        squamosa    Mollusca  
11788         algirus    Mollusca  
11798           gigas    Mollusca  

[400 rows x 6 columns]

In [4]:
#deleting empty rows
dataset.dropna(inplace=True)
dataset

class            order          family        genus  \
0         Insecta      Hymenoptera      Formicidae   Adetomyrma   
1         Insecta          Odonata       Gomphidae  Hemigomphus   
2         Insecta          Odonata  Platystictidae  Protosticta   
3         Insecta      Hymenoptera      Formicidae     Pheidole   
4         Insecta       Coleoptera    Cerambycidae  Macrodontia   
...           ...              ...             ...          ...   
11775  Gastropoda    Neogastropoda         Conidae        Conus   
11776  Gastropoda      Lepetellida      Haliotidae     Haliotis   
11777    Bivalvia         Cardiida       Cardiidae     Tridacna   
11788  Gastropoda  Stylommatophora       Zonitidae      Zonites   
11798    Bivalvia         Cardiida       Cardiidae     Tridacna   

              species      phylum  
0            venatrix  Arthropoda  
1       theischingeri  Arthropoda  
2      sanguinostigma  Arthropoda  
3            elecebra  Arthropoda  
4         cervicornis  Arthropoda  
...               ...         ...  
11775        mercator    Mollusca  
11776            roei    Mollusca  
11777        squamosa    Mollusca  
11788         algirus    Mollusca  
11798           gigas    Mollusca  

[400 rows x 6 columns]

In [5]:
#resets the index of the table to be in a numerical order 
#drops the index column with previous index value 
dataset.reset_index(drop=True, inplace=True)
dataset

class            order          family        genus         species  \
0       Insecta      Hymenoptera      Formicidae   Adetomyrma        venatrix   
1       Insecta          Odonata       Gomphidae  Hemigomphus   theischingeri   
2       Insecta          Odonata  Platystictidae  Protosticta  sanguinostigma   
3       Insecta      Hymenoptera      Formicidae     Pheidole        elecebra   
4       Insecta       Coleoptera    Cerambycidae  Macrodontia     cervicornis   
..          ...              ...             ...          ...             ...   
395  Gastropoda    Neogastropoda         Conidae        Conus        mercator   
396  Gastropoda      Lepetellida      Haliotidae     Haliotis            roei   
397    Bivalvia         Cardiida       Cardiidae     Tridacna        squamosa   
398  Gastropoda  Stylommatophora       Zonitidae      Zonites         algirus   
399    Bivalvia         Cardiida       Cardiidae     Tridacna           gigas   

         phylum  
0    Arthropoda  
1    Arthropoda  
2    Arthropoda  
3    Arthropoda  
4    Arthropoda  
..          ...  
395    Mollusca  
396    Mollusca  
397    Mollusca  
398    Mollusca  
399    Mollusca  

[400 rows x 6 columns]

In [6]:
column_entries=dataset['phylum'].value_counts()
column_entries

phylum
Chordata         332
Arthropoda        32
Cnidaria          27
Mollusca           7
Echinodermata      2
Name: count, dtype: int64

In [7]:
chordata=dataset[dataset['phylum']=='Chordata']
chordata

class               order            family         genus  \
32         Reptilia          Testudines       Geoemydidae     Pangshura   
33         Amphibia             Caudata  Cryptobranchidae       Andrias   
34             Aves       Passeriformes          Turdidae     Myadestes   
35         Mammalia            Primates         Indriidae   Propithecus   
36             Aves       Passeriformes      Fringillidae      Drepanis   
..              ...                 ...               ...           ...   
359        Reptilia          Crocodilia     Alligatoridae  Melanosuchus   
360        Mammalia           Carnivora        Mustelidae         Aonyx   
361  Chondrichthyes   Carcharhiniformes    Carcharhinidae  Carcharhinus   
362            Aves         Apodiformes       Trochilidae     Lophornis   
363  Actinopterygii  Cyprinodontiformes         Goodeidae  Zoogoneticus   

           species    phylum  
32     sylhetensis  Chordata  
33       japonicus  Chordata  
34        obscurus  Chordata  
35        deckenii  Chordata  
36        coccinea  Chordata  
..             ...       ...  
359          niger  Chordata  
360       capensis  Chordata  
361  amblyrhynchos  Chordata  
362      chalybeus  Chordata  
363        tequila  Chordata  

[332 rows x 6 columns]

In [9]:
#adding organisms to the dataframe with Echinodermata as the phylum name
new_data=[
    ["Asteroidea", "Forcipulatida", "Asteriidae", "Asterias", "Asterias rubens", "Echinodermata"],
    ["Echinoidea", "Clypeasteroida", "Clypeasteridae", "Clypeaster", "Clypeaster subdepressus", "Echinodermata"],
    ["Ophiuroidea", "Ophiurida", "Ophiuridae", "Ophiura", "Ophiura ophiura", "Echinodermata"],
    ["Crinoidea", "Comatulida", "Antedonidae", "Antedon", "Antedon mediterranea", "Echinodermata"],
    ["Holothuroidea", "Synallactida", "Synallactidae", "Parastichopus", "Parastichopus californicus", "Echinodermata"],
    ["Echinoidea", "Arbacioida", "Arbaciidae", "Arbacia", "Arbacia punctulata", "Echinodermata"],
    ["Asteroidea", "Valvatida", "Asterinidae", "Patiria", "Patiria miniata", "Echinodermata"],
    ["Ophiuroidea", "Ophiurida", "Ophiothrichidae", "Ophiothrix", "Ophiothrix fragilis", "Echinodermata"],
    ["Holothuroidea", "Dendrochirotida", "Cucumariidae", "Cucumaria", "Cucumaria frondosa", "Echinodermata"],
    ["Asteroidea", "Spinulosida", "Asterinidae", "Asterina", "Asterina gibbosa", "Echinodermata"],
    ["Asteroidea", "Velatida", "Pterasteridae", "Pteraster", "Pteraster militaris", "Echinodermata"],
    ["Echinoidea", "Diadematoida", "Diadematidae", "Diadema", "Diadema setosum", "Echinodermata"],
    ["Ophiuroidea", "Ophiacanthida", "Ophiacanthidae", "Ophiacantha", "Ophiacantha bidentata", "Echinodermata"],
    ["Holothuroidea", "Aspidochirotida", "Holothuriidae", "Holothuria", "Holothuria edulis", "Echinodermata"],
    ["Asteroidea", "Paxillosida", "Luidiidae", "Luidia", "Luidia clathrata", "Echinodermata"],
    ["Echinoidea", "Spatangoida", "Brissidae", "Brissus", "Brissus unicolor", "Echinodermata"],
    ["Crinoidea", "Cyrtocrinida", "Bathycrinidae", "Bathycrinus", "Bathycrinus aldrichianus", "Echinodermata"],
    ["Ophiuroidea", "Euryalida", "Gorgonocephalidae", "Gorgonocephalus", "Gorgonocephalus eucnemis", "Echinodermata"]
]

new_dataset =pd.DataFrame(new_data, columns=dataset.columns)
dataset.iloc[32:32+len(new_dataset)] = new_dataset

dataset.to_csv('updated_dataset.csv', index=False)
dataset.iloc[32:52]

class              order             family            genus  \
32     Asteroidea      Forcipulatida         Asteriidae         Asterias   
33     Echinoidea     Clypeasteroida     Clypeasteridae       Clypeaster   
34    Ophiuroidea          Ophiurida         Ophiuridae          Ophiura   
35      Crinoidea         Comatulida        Antedonidae          Antedon   
36  Holothuroidea       Synallactida      Synallactidae    Parastichopus   
37     Echinoidea         Arbacioida         Arbaciidae          Arbacia   
38     Asteroidea          Valvatida        Asterinidae          Patiria   
39    Ophiuroidea          Ophiurida    Ophiothrichidae       Ophiothrix   
40  Holothuroidea    Dendrochirotida       Cucumariidae        Cucumaria   
41     Asteroidea        Spinulosida        Asterinidae         Asterina   
42     Asteroidea           Velatida      Pterasteridae        Pteraster   
43     Echinoidea       Diadematoida       Diadematidae          Diadema   
44    Ophiuroidea      Ophiacanthida     Ophiacanthidae      Ophiacantha   
45  Holothuroidea    Aspidochirotida      Holothuriidae       Holothuria   
46     Asteroidea        Paxillosida          Luidiidae           Luidia   
47     Echinoidea        Spatangoida          Brissidae          Brissus   
48      Crinoidea       Cyrtocrinida      Bathycrinidae      Bathycrinus   
49    Ophiuroidea          Euryalida  Gorgonocephalidae  Gorgonocephalus   
50       Reptilia           Squamata         Gekkonidae         Phelsuma   
51           Aves  Procellariiformes     Procellariidae      Procellaria   

                       species         phylum  
32             Asterias rubens  Echinodermata  
33     Clypeaster subdepressus  Echinodermata  
34             Ophiura ophiura  Echinodermata  
35        Antedon mediterranea  Echinodermata  
36  Parastichopus californicus  Echinodermata  
37          Arbacia punctulata  Echinodermata  
38             Patiria miniata  Echinodermata  
39         Ophiothrix fragilis  Echinodermata  
40          Cucumaria frondosa  Echinodermata  
41            Asterina gibbosa  Echinodermata  
42         Pteraster militaris  Echinodermata  
43             Diadema setosum  Echinodermata  
44       Ophiacantha bidentata  Echinodermata  
45           Holothuria edulis  Echinodermata  
46            Luidia clathrata  Echinodermata  
47            Brissus unicolor  Echinodermata  
48    Bathycrinus aldrichianus  Echinodermata  
49    Gorgonocephalus eucnemis  Echinodermata  
50                   standingi       Chordata  
51                  parkinsoni       Chordata

In [10]:
#adding more organisms to the dataframe with Mollusca as the phylum name
new_organisms=[
    ["Bivalvia", "Ostreoida", "Ostreidae", "Crassostrea", "gigas", "Mollusca"],
    ["Bivalvia", "Veneroida", "Veneridae", "Mercenaria", "mercenaria", "Mollusca"],
    ["Gastropoda", "Pulmonata", "Limacidae", "Limax", "maximus", "Mollusca"],
    ["Gastropoda", "Neogastropoda", "Buccinidae", "Busycon", "carica", "Mollusca"],
    ["Bivalvia", "Myida", "Myidae", "Mytilus", "edulis", "Mollusca"],
    ["Cephalopoda", "Octopoda", "Octopodidae", "Octopus", "vulgaris", "Mollusca"],
    ["Gastropoda", "Heterobranchia", "Stylommatophora", "Eobrachycera", "muscosa", "Mollusca"],
    ["Bivalvia", "Pteriida", "Pteriidae", "Pinctada", "margaritifera", "Mollusca"],
    ["Gastropoda", "Caenogastropoda", "Fasciolariidae", "Fasciolaria", "tulipa", "Mollusca"],
    ["Gastropoda", "Caenogastropoda", "Cypraeidae", "Cypraea", "tigris", "Mollusca"],
    ["Cephalopoda", "Decapodiformes", "Loliginidae", "Loligo", "vulgaris", "Mollusca"],
    ["Gastropoda", "Cerithioidea", "Thiaridae", "Thiara", "granifera", "Mollusca"],
    ["Bivalvia", "Veneroida", "Donacidae", "Donax", "variabilis", "Mollusca"],
    ["Gastropoda", "Basommatophora", "Planorbidae", "Planorbarius", "corneus", "Mollusca"],
    ["Cephalopoda", "Nautilida", "Nautilidae", "Nautilus", "pompilius", "Mollusca"],
    ["Bivalvia", "Pectinida", "Pectinidae", "Pecten", "maximus", "Mollusca"],
    ["Gastropoda", "Opisthobranchia", "Aplysiidae", "Aplysia", "californica", "Mollusca"],
    ["Bivalvia", "Cardiida", "Cardiidae", "Cardium", "edule", "Mollusca"],
    ["Gastropoda", "Caenogastropoda", "Neritidae", "Nerita", "atramentosa", "Mollusca"]
]
# Adding the new organisms starting from index 50
dataset.iloc[50:50+len(new_organisms), :] = new_organisms
dataset.to_csv('updated_dataset.csv', index=False)


dataset.iloc[50:70]

class            order           family         genus  \
50     Bivalvia        Ostreoida        Ostreidae   Crassostrea   
51     Bivalvia        Veneroida        Veneridae    Mercenaria   
52   Gastropoda        Pulmonata        Limacidae         Limax   
53   Gastropoda    Neogastropoda       Buccinidae       Busycon   
54     Bivalvia            Myida           Myidae       Mytilus   
55  Cephalopoda         Octopoda      Octopodidae       Octopus   
56   Gastropoda   Heterobranchia  Stylommatophora  Eobrachycera   
57     Bivalvia         Pteriida        Pteriidae      Pinctada   
58   Gastropoda  Caenogastropoda   Fasciolariidae   Fasciolaria   
59   Gastropoda  Caenogastropoda       Cypraeidae       Cypraea   
60  Cephalopoda   Decapodiformes      Loliginidae        Loligo   
61   Gastropoda     Cerithioidea        Thiaridae        Thiara   
62     Bivalvia        Veneroida        Donacidae         Donax   
63   Gastropoda   Basommatophora      Planorbidae  Planorbarius   
64  Cephalopoda        Nautilida       Nautilidae      Nautilus   
65     Bivalvia        Pectinida       Pectinidae        Pecten   
66   Gastropoda  Opisthobranchia       Aplysiidae       Aplysia   
67     Bivalvia         Cardiida        Cardiidae       Cardium   
68   Gastropoda  Caenogastropoda        Neritidae        Nerita   
69         Aves     Anseriformes         Anatidae     Somateria   

          species    phylum  
50          gigas  Mollusca  
51     mercenaria  Mollusca  
52        maximus  Mollusca  
53         carica  Mollusca  
54         edulis  Mollusca  
55       vulgaris  Mollusca  
56        muscosa  Mollusca  
57  margaritifera  Mollusca  
58         tulipa  Mollusca  
59         tigris  Mollusca  
60       vulgaris  Mollusca  
61      granifera  Mollusca  
62     variabilis  Mollusca  
63        corneus  Mollusca  
64      pompilius  Mollusca  
65        maximus  Mollusca  
66    californica  Mollusca  
67          edule  Mollusca  
68    atramentosa  Mollusca  
69       fischeri  Chordata

In [11]:
#shuffling the dataset and returning all (100%) of the rows but shuffled 
dataset= dataset.sample(frac=1)
dataset

class              order             family          genus  \
54         Bivalvia              Myida             Myidae        Mytilus   
332        Mammalia           Primates    Cercopithecidae  Semnopithecus   
80             Aves        Otidiformes           Otididae      Eupodotis   
151        Amphibia              Anura     Arthroleptidae     Leptopelis   
313        Mammalia           Primates          Hominidae          Pongo   
..              ...                ...                ...            ...   
74             Aves      Passeriformes      Bombycillidae     Bombycilla   
309  Actinopterygii  Chaetodontiformes     Chaetodontidae      Chaetodon   
17          Insecta           Phasmida  Pseudophasmatidae     Peruphasma   
0           Insecta        Hymenoptera         Formicidae     Adetomyrma   
239  Chondrichthyes  Carcharhiniformes     Carcharhinidae   Carcharhinus   

          species      phylum  
54         edulis    Mollusca  
332        johnii    Chordata  
80   caerulescens    Chordata  
151  vermiculatus    Chordata  
313        abelii    Chordata  
..            ...         ...  
74       japonica    Chordata  
309  trifascialis    Chordata  
17       schultei  Arthropoda  
0        venatrix  Arthropoda  
239  melanopterus    Chordata  

[400 rows x 6 columns]

In [12]:
#renumbering the row indices
dataset.reset_index(drop=True, inplace=True)
dataset

class              order             family          genus  \
0          Bivalvia              Myida             Myidae        Mytilus   
1          Mammalia           Primates    Cercopithecidae  Semnopithecus   
2              Aves        Otidiformes           Otididae      Eupodotis   
3          Amphibia              Anura     Arthroleptidae     Leptopelis   
4          Mammalia           Primates          Hominidae          Pongo   
..              ...                ...                ...            ...   
395            Aves      Passeriformes      Bombycillidae     Bombycilla   
396  Actinopterygii  Chaetodontiformes     Chaetodontidae      Chaetodon   
397         Insecta           Phasmida  Pseudophasmatidae     Peruphasma   
398         Insecta        Hymenoptera         Formicidae     Adetomyrma   
399  Chondrichthyes  Carcharhiniformes     Carcharhinidae   Carcharhinus   

          species      phylum  
0          edulis    Mollusca  
1          johnii    Chordata  
2    caerulescens    Chordata  
3    vermiculatus    Chordata  
4          abelii    Chordata  
..            ...         ...  
395      japonica    Chordata  
396  trifascialis    Chordata  
397      schultei  Arthropoda  
398      venatrix  Arthropoda  
399  melanopterus    Chordata  

[400 rows x 6 columns]

In [13]:
dataset.tail(10)

class              order             family           genus  \
390  Chondrichthyes        Lamniformes       Cetorhinidae      Cetorhinus   
391        Anthozoa       Scleractinia       Euphylliidae  Fimbriaphyllia   
392            Aves    Charadriiformes         Burhinidae          Esacus   
393        Mammalia           Primates            Cebidae           Cebus   
394        Mammalia            Cetacea        Delphinidae           Sousa   
395            Aves      Passeriformes      Bombycillidae      Bombycilla   
396  Actinopterygii  Chaetodontiformes     Chaetodontidae       Chaetodon   
397         Insecta           Phasmida  Pseudophasmatidae      Peruphasma   
398         Insecta        Hymenoptera         Formicidae      Adetomyrma   
399  Chondrichthyes  Carcharhiniformes     Carcharhinidae    Carcharhinus   

          species      phylum  
390       maximus    Chordata  
391        ancora    Cnidaria  
392  magnirostris    Chordata  
393      unicolor    Chordata  
394       plumbea    Chordata  
395      japonica    Chordata  
396  trifascialis    Chordata  
397      schultei  Arthropoda  
398      venatrix  Arthropoda  
399  melanopterus    Chordata

In [14]:
from sklearn.preprocessing import LabelEncoder
import pandas as pd


In [15]:
# Apply Label Encoding to all categorical columns
label_encoder = LabelEncoder()

In [16]:
 #Identifying columns that need encoding
categorical_columns = ['class', 'order', 'family', 'genus', 'species',]

In [17]:
# Encoding each categorical column
for column in categorical_columns:
    dataset[column] = label_encoder.fit_transform(dataset[column])

In [18]:
dataset

class  order  family  genus  species      phylum
0        6     56     131    197      126    Mollusca
1       16     80      44    282      183    Chordata
2        5     69     143    118       72    Chordata
3        1      4      17    167      365    Chordata
4       16     80     103    249       18    Chordata
..     ...    ...     ...    ...      ...         ...
395      5     70      27     47      178    Chordata
396      0     19      47     77      354    Chordata
397     14     77     167    235      317  Arthropoda
398     14     48      88      6      364  Arthropoda
399      9     13      39     64      229    Chordata

[400 rows x 6 columns]

In [19]:
from sklearn.model_selection import train_test_split, KFold, cross_val_score
from sklearn import model_selection

In [20]:
array = dataset.values
X = array[:, 0:5]
Y = array[:, 5]
validation_size = 0.20
seed = 7
X_train, X_validation, Y_train, Y_validation =model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)

In [21]:
seed = 42
scoring = 'accuracy'

In [22]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression

In [33]:
models = []
models.append(('LR', LogisticRegression()))
models.append(('LDA', LinearDiscriminantAnalysis()))
models.append(('KNN', KNeighborsClassifier()))
models.append(('CART', DecisionTreeClassifier()))
models.append(('NB', GaussianNB()))
models.append(('SVM', SVC()))
# print(models)
# evaluate each model in turn
results = []
names = []
for name, model in models:
        kfold = model_selection.KFold(n_splits=10, random_state=None)
        cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring)
        results.append(cv_results)
        names.append(name)
        msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
        print(msg)

LR: 0.828125 (0.072954)
LDA: 0.790625 (0.099853)
KNN: 0.753125 (0.089976)
CART: 0.959375 (0.028125)
NB: 0.850000 (0.065252)
SVM: 0.759375 (0.099853)


C:\Users\TECH\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning:

lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression

C:\Users\TECH\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning:

lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please

visualization

In [24]:
# Top 10 Families by Species Count
import plotly.express as px
top_families = dataset['family'].value_counts().nlargest(10).reset_index()
top_families.columns = ['family', 'count']

fig = px.bar(top_families, x='count', y='family', orientation='h', color='family',
             title='Families by Species Count')
fig.show()


In [25]:
# Species count by Class
class_count = dataset['class'].value_counts().reset_index()
class_count.columns = ['class', 'count']

fig = px.bar(class_count, x='class', y='count', color='class',
             title='Histogram species Count by Class')
fig.show()


In [26]:



phylum_count = dataset['phylum'].value_counts().reset_index()
phylum_count.columns = ['phylum', 'phylum_count']


class_count = dataset['class'].value_counts().reset_index()
class_count.columns = ['class', 'class_count']


scatter_data = phylum_count.merge(class_count, how='cross')  


fig = px.scatter(scatter_data, x='phylum_count', y='class_count',
                 title='Scatter Plot of Species Count by Class and Phylum',
                 labels={'phylum_count': 'Phylum Count', 'class_count': 'Class Count'},
                 hover_data=['phylum', 'class'])  
fig.show()


In [27]:
import plotly.express as px

# Create a dataframe for phylum counts
phylum_count = dataset['phylum'].value_counts().reset_index()
phylum_count.columns = ['phylum', 'count']


fig_phylum = px.bar(phylum_count, x='phylum', y='count',
                    title='Species Count by Phylum',
                    labels={'phylum': 'Phylum', 'count': 'Species Count'},
                    color='phylum')  
fig_phylum.show()


In [29]:
import plotly.figure_factory as ff
# Extract the 'class' column
class_data = dataset['class']

# Create a distribution plot (histogram + normal distribution curve)
fig = ff.create_distplot([class_data], ['Class Data'], bin_size=0.5, show_rug=False)

# Add a title
fig.update_layout(title='Normal Distribution of Class Column')

# Show the plot
fig.show()

In [38]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.metrics import classification_report, confusion_matrix


DTC = DecisionTreeClassifier()
DTC.fit(X_train, Y_train)
predictions = DTC.predict(X_validation)
print(accuracy_score(Y_validation, predictions))
print("\t=============================================\n")
print(confusion_matrix(Y_validation, predictions))
print("\t=============================================\n")
print(classification_report(Y_validation, predictions))

0.9625

[[ 8  1  0  0  0]
 [ 0 52  0  0  0]
 [ 0  0  8  0  0]
 [ 0  0  0  4  0]
 [ 0  0  0  2  5]]

               precision    recall  f1-score   support

   Arthropoda       1.00      0.89      0.94         9
     Chordata       0.98      1.00      0.99        52
     Cnidaria       1.00      1.00      1.00         8
Echinodermata       0.67      1.00      0.80         4
     Mollusca       1.00      0.71      0.83         7

     accuracy                           0.96        80
    macro avg       0.93      0.92      0.91        80
 weighted avg       0.97      0.96      0.96        80



For the DecisionTreeClassifier model achieved an accuracy of 96.25%. 

Furthermore from the confusion matrix there were only 3 incorrect predictions.

Then proceeding to the classification report, out of all predictions made for the Arthropoda,Chordata,Chidaria,Echinodermata and Mollusca , 100%,98%,100%,67% and 100% were correct respectively. 

Out of all actual Arthropoda,Chordata,Chidaria,Echinodermata and Mollusca ,89%,100%,100%,100% and 71% respectively were classified correcctly. 

in F1-Score we see that the balance between precision and recall is very good throughout.

From support we see that there were 9,52,8,4 and 7 actual instances of Arthropoda,Chordata,Chidaria,Echinodermata and Mollusca.

The overall accuracy of the model is 96%
The unweighted average of precison,recall and f1-score is 93%,92% and 91%. This shows that its handling other kinds of phylum very well actually whether minority or majority
The weighted average of precison,recall and f1-score is 97%,96% and 96% is great news especially for the fields that have more instances

In [39]:
GNB = GaussianNB()
GNB.fit(X_train, Y_train)
predictions = GNB.predict(X_validation)
print(accuracy_score(Y_validation, predictions))
print("\t=============================================\n")
print(confusion_matrix(Y_validation, predictions))
print("\t=============================================\n")
print(classification_report(Y_validation, predictions))

0.775

[[ 0  9  0  0  0]
 [ 0 51  0  1  0]
 [ 0  0  8  0  0]
 [ 0  1  0  3  0]
 [ 0  7  0  0  0]]

               precision    recall  f1-score   support

   Arthropoda       0.00      0.00      0.00         9
     Chordata       0.75      0.98      0.85        52
     Cnidaria       1.00      1.00      1.00         8
Echinodermata       0.75      0.75      0.75         4
     Mollusca       0.00      0.00      0.00         7

     accuracy                           0.78        80
    macro avg       0.50      0.55      0.52        80
 weighted avg       0.62      0.78      0.69        80



C:\Users\TECH\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning:

Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.

C:\Users\TECH\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning:

Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.

C:\Users\TECH\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning:

Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` param

For the GaussianNB model achieved an accuracy of 77.50% Which is already worse than that of the DecisionTreeClassifier model

Furthermore from the confusion matrix there were a lot of incorrect predictions. we have the model making 18 wrong preditions.This is worse than its competitor

Then proceeding to the classification report, out of all predictions made for the Arthropoda,Chordata,Chidaria,Echinodermata and Mollusca , 0%,75%,100%,75% and 0% were correct respectively. 

Out of all actual Arthropoda,Chordata,Chidaria,Echinodermata and Mollusca ,0%,98%,100%,75% and 0% respectively were classified correcctly. 

in F1-Score we see that the balance between precision and recall is not so good as compared to the DecisionTreeClassifier model as it lacks consistency in results

From support we see that there were 9,52,8,4 and 7 actual instances of Arthropoda,Chordata,Chidaria,Echinodermata and Mollusca.

The unweighted average of precison,recall and f1-score show that this model is handling other kinds of phylum very badly whether minority or majority compared to the DecisionTreeClassifier model
The weighted average of precison,recall and f1-score also shows that it is not great especially for the fields that have more instances

In [41]:
from sklearn import metrics
print(metrics.accuracy_score(Y, DTC.predict(X)))
print(metrics.accuracy_score(Y, GNB.predict(X)))

0.9925
0.8475


Then to finally hammer the final nail, the accuracy was tested using x and y values and to no ones suprise, the DecisionTreeClassifier model had much higher accuracy score of 99.25% making this model the ideal choice for the problem